src/adapters/ip_camera/mod.rs

```diff
@@ -163,11 +164,123 @@ impl IPCameraAdapter {
+            id: getter_location_id.clone(),
+            last_seen: None,
+            service: service_id.clone(),
+            mechanism: Getter {
+                kind: ChannelKind::Location,
+                updated: None,
+            },
+        }));
+
+        let setter_location_id = create_setter_id("location", udn);
+        try!(adapt.add_setter(Channel {
+            tags: HashSet::new(),
+            adapter: adapter_id.clone(),
+            id: setter_location_id.clone(),
+            last_seen: None,
+            service: service_id.clone(),
+            mechanism: Setter {
+                kind: ChannelKind::Location,
+                updated: None,
+            },
+        }));
+
+        let getter_username_id = create_getter_id("username", udn);
+        try!(adapt.add_getter(Channel {
+            tags: HashSet::new(),
+            adapter: adapter_id.clone(),
+            id: getter_username_id.clone(),
+            last_seen: None,
+            service: service_id.clone(),
+            mechanism: Getter {
+                kind: ChannelKind::Username,
+                updated: None,
+            },
+        }));
+
+        let setter_username_id = create_setter_id("username", udn);
+        try!(adapt.add_setter(Channel {
+            tags: HashSet::new(),
+            adapter: adapter_id.clone(),
+            id: setter_username_id.clone(),
+            last_seen: None,
+            service: service_id.clone(),
+            mechanism: Setter {
+                kind: ChannelKind::Username,
+                updated: None,
+            },
+        }));
+
+        let getter_password_id = create_getter_id("password", udn);
+        try!(adapt.add_getter(Channel {
+            tags: HashSet::new(),
+            adapter: adapter_id.clone(),
+            id: getter_password_id.clone(),
+            last_seen: None,
+            service: service_id.clone(),
+            mechanism: Getter {
+                kind: ChannelKind::Password,
+                updated: None,
+            },
+        }));
+
+        let setter_password_id = create_setter_id("password", udn);
+        try!(adapt.add_setter(Channel {
+            tags: HashSet::new(),
+            adapter: adapter_id.clone(),
+            id: setter_password_id.clone(),
+            last_seen: None,
+            service: service_id.clone(),
+            mechanism: Setter {
+                kind: ChannelKind::Password,
+                updated: None,
+            },
+        }));
+
         let mut serv = services.lock().unwrap();
-        let camera = try!(IpCamera::new(udn, url, &serv.snapshot_root));
+        let camera = try!(IpCamera::new(udn, url, name, &serv.snapshot_root, config));
         serv.getters.insert(getter_image_list_id, camera.clone());
         serv.getters.insert(getter_image_newest_id, camera.clone());
-        serv.setters.insert(setter_snapshot_id, camera);
+        serv.setters.insert(setter_snapshot_id, camera.clone());
```

Comment:
Having so many clones of the camera seems a bit weird. I would probably have used an `Arc`. YMMV.